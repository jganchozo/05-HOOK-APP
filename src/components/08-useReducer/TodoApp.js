import React, { useEffect, useReducer } from 'react'
import './styles.css';
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';


const init = () => {

    return JSON.parse(localStorage.getItem('todos')) || [];

    // return [{
    //     id: new Date().getTime(),
    //     desc: 'Aprender React',
    //     done: false,
    // }];
};

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
      localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);


    const handleDelete = (id) => {

        const action = {
            type: 'delete',
            payload: id,
        };

        dispatch(action);
    };

    const handleToggle = (id) => {
        dispatch({
            type: 'toggle',
            payload: id,
        });
    };

    const handleAddTodo = (newTodo) => {

        dispatch({
            type: 'add',
            payload: newTodo,
        });

    };

    return (
        <>
            <h1>TodoApp ( {todos.length} )</h1>
            <hr />

            <div className='row'>
                <div className='col-7'>
                    <TodoList
                        todos={todos}
                        handleDelete={handleDelete}
                        handleToggle={handleToggle}
                    />
                </div>
                <div className='col5'>
                    <TodoAdd handleAddTodo={handleAddTodo} />
                </div>

            </div>
        </>
    )
}