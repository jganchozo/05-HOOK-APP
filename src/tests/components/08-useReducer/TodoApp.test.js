import { act } from '@testing-library/react';
import { mount, shallow } from 'enzyme';
import React from 'react';
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { demoTodos } from '../../ficxtures/demoTodos';
import '@testing-library/jest-dom';

describe('Pruebas en <TodoApp />', () => { 
    
    const wrapper = shallow(<TodoApp />);

    Storage.prototype.setItem = jest.fn(() => {});

    
    test('should de mostrarse correctamente', () => {

        expect(wrapper).toMatchSnapshot();
    });

    test('should de agregar un todo', () => { 

        const wrapper = mount(<TodoApp />);
        act(() => {
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1]);
        });

        expect(wrapper.find('h1').text().trim()).toBe(`TodoApp ( ${demoTodos.length} )`);
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
        //expect(localStorage.setItem).toHaveBeenCalledWith({});

    });

    test('should de eliminar un todo', () => { 
       
        wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
        wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);

        expect(wrapper.find('h1').text().trim()).toBe(`TodoApp ( ${0} )`);
    });

 });