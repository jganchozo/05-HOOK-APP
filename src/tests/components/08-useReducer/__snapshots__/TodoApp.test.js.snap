// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp /> should de mostrarse correctamente 1`] = `
<Fragment>
  <h1>
    TodoApp ( 
    0
     )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
